plugin/src/HeadlessAPI.lua

```diff
@@ -0,0 +1,495 @@
+		Rojo._activePermissionRequests[source] = true
+
+		-- Sanitize request
+		local sanitizedApis = {}
+		for _, api in apis do
+			if Rojo._apiDescriptions[api] ~= nil and table.find(sanitizedApis, api) == nil then
+				table.insert(sanitizedApis, api)
+			else
+				warn(string.format("Rojo.%s is not a valid API", tostring(api)))
+			end
+		end
+		assert(#sanitizedApis > 0, "Rojo:RequestAccess expects an array of valid API names")
+		table.sort(sanitizedApis)
+
+		local alreadyAllowed = true
+		if Rojo._permissions[source] == nil then
+			alreadyAllowed = false
+		else
+			for _, api in sanitizedApis do
+				if not Rojo._permissions[source][api] then
+					alreadyAllowed = false
+					break
+				end
+			end
+		end
+
+		if alreadyAllowed then
+			Rojo._activePermissionRequests[source] = nil
+			return true
+		end
+
+		local granted = app:requestPermission(plugin, source, name, sanitizedApis, false)
+		if granted then
+			Rojo:_setPermissions(source, name, sanitizedApis)
+		else
+			Rojo:_removePermissions(source, name)
+		end
+
+		Rojo._activePermissionRequests[source] = nil
+		return granted
+	end
+
+	Rojo._apiDescriptions.Test = {
+		Type = "Method",
+		Description = "Prints the given arguments to the console. Useful during development for testing purposes.",
+	}
+	function Rojo:Test(...)
+		local args = table.pack(...)
+		for i = 1, args.n do
+			local v = args[i]
+			local t = type(v)
+			if t == "string" then
+				args[i] = string.format("%q", v)
+			else
+				args[i] = tostring(v)
+			end
+		end
+
+		print(
+			string.format(
+				"Rojo:Test(%s) called from '%s' (%s)",
+				table.concat(args, ", "),
+				Rojo:_getCallerName(),
+				Rojo:_getCallerSource()
+			)
+		)
+	end
+
+	Rojo._apiDescriptions.ConnectAsync = {
+		Type = "Method",
+		Description = "Connects to a Rojo server",
+	}
+	function Rojo:ConnectAsync(host: string?, port: string?)
+		assert(type(host) == "string" or host == nil, "Host must be type `string?`")
+		assert(type(port) == "string" or port == nil, "Port must be type `string?`")
+
+		if Rojo:_checkRateLimit("ConnectAsync") then
+			return
+		end
+
+		app:startSession(host, port)
+	end
+
+	Rojo._apiDescriptions.DisconnectAsync = {
+		Type = "Method",
+		Description = "Disconnects from the Rojo server",
+	}
+	function Rojo:DisconnectAsync()
+		if Rojo:_checkRateLimit("DisconnectAsync") then
+			return
+		end
+
+		app:endSession()
+	end
+
+	Rojo._apiDescriptions.GetSetting = {
+		Type = "Method",
+		Description = "Gets a Rojo setting",
+	}
+	function Rojo:GetSetting(setting: string): any
+		assert(type(setting) == "string", "Setting must be type `string`")
+
+		return Settings:get(setting)
+	end
+
+	Rojo._apiDescriptions.SetSetting = {
+		Type = "Method",
+		Description = "Sets a Rojo setting",
+	}
+	function Rojo:SetSetting(setting: string, value: any)
+		assert(type(setting) == "string", "Setting must be type `string`")
+		assert(type(value) == type(Settings:get(setting)), "Value must be the same type as the setting")
```

Comment:
This only checks the types but we should probably check the values as well. For example `confirmationBehavior` is a `string` but it only has a few possible values.

Reply:
The `t` library would be great for this and I think we already use it as a dependency.